### PR TITLE
My proposal for generating the client cert and key.

### PR DESCRIPTION
**Pull request by mbruzek**
Hey Chuck,

I need the client key for kubernetes too so I took a stab at generating the client key.  This is ONLY a proposal and not tested or verified yet!  I just wanted to get something in by the EOD so we could discuss approaches.

Please note that this changes the state name that other layers can consume to 'tls.server.certificate available' or 'tls.client.certificate available'.

Please take a look and comment about what you like or do not like.

### CODE REVIEW

File: reactive/tls.py [hunk 1]
```diff
@@ -187,16 +187,29 @@ def create_server_certificate():
...
+        # Do not regenerate the client certificate if it already exists.
+        if not os.path.isfile(client_file):
+            # Create a client certificate and key.
+            client = './easyrsa --batch --req-cn={0} --subject-alt-name={1} ' \
+                     'build-client-full client nopass 2>&1'.format(cn, sans)
+            check_call(split(client))
+            # Read the client certificate from the filesystem.
+            with open(client_file, 'r') as fp:
+                client_cert = fp.read()
+            set_cert('tls.client.certificate', client_cert)
```
Comment (chuckbutler) [inline reactive/tls.py]: this will fail idempotency checks :(

File: reactive/tls.py [hunk 2]
```diff
@@ -133,18 +133,18 @@ def copy_server_cert(tls):
...
     # Set cert on the unitdata key value store so other layers can get it.
-    unitdata.kv().set('tls.server.certificate', server_cert)
+    unitdata.kv().set(key, certificate)
     # Set the final state for the other layers to know when they can
     # retrieve the server certificate.
-    set_state('server certificate available')
+    set_state('{0} available'.format(key))
 
 
 @when_not('certificate authority available')
```
Comment (chuckbutler) [inline reactive/tls.py]: This will also fail if the CA is not generated prior to executing the client certificate bits

File: reactive/tls.py [hunk 3]
```diff
@@ -133,18 +133,18 @@ def copy_server_cert(tls):
...
     if cert:
-        set_server_cert(cert)
+        set_cert('tls.server.certificate', cert)
         remove_state('signed certificate available')
 
 
-def set_server_cert(server_cert):
-    '''Set the serve certificate on the key value store of the unit, and set
-    the final state for layers to consume.'''
+def set_cert(key, certificate):
```
Comment (chuckbutler) [inline reactive/tls.py]: I like this idea of extracting out the KV storage of the keys

File: reactive/tls.py (hunk 1 above)
Comment (mbruzek) [inline reactive/tls.py]: The guard on 204 should prevent it from running again, I even have a comment!

File: reactive/tls.py (hunk 2 above)
Comment (mbruzek) [inline reactive/tls.py]: The client and server certificate are generated together.  The server certificate never was called before the ca was generated which is the reason for this proposal, I wanted to discuss approach

File: reactive/tls.py (hunk 3 above)
Comment (mbruzek) [inline reactive/tls.py]: Yes this method involves setting the kv store, and the reactive state, I thought it was a good example of reuse.

File: reactive/tls.py [hunk 4]
```diff
@@ -181,22 +181,35 @@ def create_certificate_authority(certificate_authority=None):
...
     with chdir('easy-rsa/easyrsa3'):
-        server_file = 'pki/issued/{0}.crt'.format(cn)
+        server_file = 'pki/issued/server.crt'
+        sans = get_sans()
         # Do not regenerate the server certificate if it already exists.
         if not os.path.isfile(server_file):
             # Create a server certificate for the server based on the CN.
             server = './easyrsa --batch --req-cn={0} --subject-alt-name={1} ' \
-                     'build-server-full {0} nopass 2>&1'.format(cn, get_sans())
+                     'build-server-full server nopass 2>&1'.format(cn, sans)
```
Comment (mbruzek) [inline reactive/tls.py]: This change makes the server certificate go from PUBLIC_IP.crt to "server.crt" which is like the other units (so we have less special case code).  If this is a problem we should revert this change.

**Comment by mbruzek** [comment]
Thanks for the review @chuckbutler!  There are some controversial changes here, and obviously different than your approach.  I would like to discuss the approaches and pick the best one ... even if it is not this one.

File: reactive/tls.py [hunk 5]
```diff
@@ -127,28 +127,29 @@ def import_sign(tls):
...
+def set_cert(key, certificate):
+    '''Set the certificate on the key value store of the unit, and set
+    the corresponding state for layers to consume.'''
     # Set cert on the unitdata key value store so other layers can get it.
-    unitdata.kv().set('tls.server.certificate', server_cert)
+    unitdata.kv().set(key, certificate)
     # Set the final state for the other layers to know when they can
     # retrieve the server certificate.
-    set_state('server certificate available')
+    set_state('{0} available'.format(key))
```
Comment (chuckbutler) [inline reactive/tls.py]: yeah buddy!

File: reactive/tls.py [hunk 6]
```diff
@@ -181,22 +182,37 @@ def create_certificate_authority(certificate_authority=None):
...
 
 
-def create_server_certificate():
+def create_certificates():
     '''Create the server certificate.'''
     # Use the public ip as the Common Name for the server certificate.
     cn = hookenv.unit_public_ip()
     with chdir('easy-rsa/easyrsa3'):
-        server_file = 'pki/issued/{0}.crt'.format(cn)
+        server_file = 'pki/issued/server.crt'
```
Comment (chuckbutler) [inline reactive/tls.py]: i really think this should be the sanitized unit id.

```
server = getenv('JUJU_UNIT_NAME').replace('/','_')
server_file = 'pki/issued/{}.crt'.format(server)
```

File: reactive/tls.py [hunk 7]
```diff
@@ -208,6 +224,8 @@ def install_ca(certificate_authority):
         fp.write(certificate_authority)
     # Update the trusted CAs on this system.
     check_call(['update-ca-certificates'])
+    # Notify other layers that the certificate authority is available.
+    set_state('tls.certificate.authority available')
```
Comment (chuckbutler) [inline reactive/tls.py]: this seems unnatural with a guard against the older format on line 152

File: reactive/tls.py [hunk 8]
```diff
@@ -127,28 +127,29 @@ def import_sign(tls):
 
 
 @when('signed certificate available')
-@when_not('server certificate available')
```
Comment (chuckbutler) [inline reactive/tls.py]: why'd this one get expunged? i thought we would want to prevent the server cert from being regenerated.

Comment (chuckbutler) [inline reactive/tls.py]: oh i see the control flow moved from a stacked qualifier to having 'signed certificate available' removed on line 138

File: reactive/tls.py (hunk 6 above)
Comment (mbruzek) [inline reactive/tls.py]: I disagree.  This change is only for the leader's server cert, all the other followers get their cert name "server.crt" so this makes all the units the same (rather than having special case code just for the leader).  I don't want to special case the leader's certs when we don't need to.  The client stuff is named "client.crt" and "client.key".

File: reactive/tls.py (hunk 4 above)
Comment (mbruzek) [inline reactive/tls.py]: It has not shown to be a problem in my testing.  I vote to keep this in so the tests and consuming layers will not have to do special cases to get the server certificate and keys.

File: reactive/tls.py (hunk 8 above)
Comment (mbruzek) [inline reactive/tls.py]: Good catch!  I used the state "server certificate available" before we knew about needing client certificates as well.  This logic was changed to be more consistent between the server and client certificate states.

File: reactive/tls.py [hunk 9]
```diff
@@ -50,7 +50,7 @@ def check_ca_status():
             install_ca(certificate_authority)
             # The leader can create the server certificate based on CA.
             hookenv.log('Leader is creating server certificate.')
-            create_server_certificate()
+            create_certificates()
```
Comment (chuckbutler) [inline reactive/tls.py]: pendatry :cop:

File: reactive/tls.py [hunk 10]
```diff
@@ -100,11 +106,12 @@ def import_sign(tls):
         csr_map = tls.get_csr_map()
         # Iterate over the unit names related to CSRs.
         for unit_name, csr in csr_map.items():
-            path_name = _path_safe_name(unit_name)
             with chdir('easy-rsa/easyrsa3'):
                 temp_file = tempfile.NamedTemporaryFile(suffix='.csr')
                 with open(temp_file.name, 'w') as fp:
                     fp.write(csr)
+                # Must remove the path characters from the unit_name.
+                path_name = unit_name.replace('/', '_')
```
Comment (chuckbutler) [inline reactive/tls.py]: this little thing... and all its annoyances. :+1: good job on eliminating that one

**Comment by chuckbutler** [comment]
+1 LGTM

Merged and happy dance ensues
